unidist/cli/base/utils.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (C) 2021 Modin authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+"""Utilities for backend runners."""
+
+import os
+import ipaddress
+import socket
+import multiprocessing as mp
+
+
+class BackendName:
+    """String representations of unidist backends."""
+
+    RAY = "ray"
+    MPI = "mpi"
+    DASK = "dask"
+    MP = "multiprocessing"
+    PY = "python"
+
+
+class Defaults:
+    """Default values for supported CLI parameters."""
+
+    EXECUTOR = "python"
+    BACKEND = BackendName.RAY
+    NUM_CPUS = "default"
+    HOSTS = "localhost"
+    REDIS_PASSWORD = "5241590000000000"
+
+
+def get_unidist_root():
+    """
+    Get the project root directory.
+
+    Returns
+    -------
+    str
+        Absolute path to the project root directory.
+    """
+    unidist_root = os.path.abspath(os.path.join(os.path.dirname(__file__), "../../../"))
+    os.environ["PYTHONPATH"] = (
+        unidist_root + os.pathsep + os.environ.get("PYTHONPATH", "")
+    )
+
+    return unidist_root
+
+
+def validate_hosts(hosts: list):
+    """
+    Validate `hosts` list of ip-addresses on correctness and duplicates.
+
+    Parameters
+    ----------
+    hosts : list
+        List of strings with ip-addresses.
+
+    Returns
+    -------
+    list
+        List of validated IPs.
+    """
+    ips = [
+        str(ipaddress.ip_address(get_localhost_ip() if ip == "localhost" else ip))
+        for ip in hosts
+    ]
+    duplicate_ips = [ip for ip in set(ips) if ips.count(ip) > 1]
+    if len(duplicate_ips):
+        raise RuntimeError(f"'hosts' list contains duplicates {duplicate_ips}")
+    return ips
+
+
+def validate_num_cpus(num_cpus: list):
+    """
+    Validate `num_cpus` on correctness.
+
+    Each value of `num_cpus` is checked on possibility
+    of converting to int. In case value is ``default`` it will
+    be equal to number of CPUs on ``localhost`` node.
+
+    Parameters
+    ----------
+    num_cpus : list
+        List of string values. The each value represents
+        a number of workers for corresponded host.
+
+    Returns
+    -------
+    list
+        List of validated number of workers per host.
+    """
+
+    def validate(value):
+        try:
+            value = int(value)
+            if value < 1:
+                raise RuntimeError(f"'num_cpus' must be more than 0, got '{num_cpus}'")
+        except ValueError:
+            if value == "default":
+                return mp.cpu_count()
```

Comment:
Why don't we use our `CpuCount.get()` here?

Reply:
In case using of `CpuCount.get()` we rely on environment variable `UNIDIST_CPUS`. If this variable will be presented in user's env we get unexpected behavior in CLI in case  `num_cpus == "default"`. In that case `num_cpus` will be set to `UNIDIST_CPUS` value, instead of `mp.cpu_count()`. But I would prefer to set up of CLI running behavior via command line parameters only.

Reply:
Then, we will have our env vars useless. We should probably have some handling/clarifying description for `num_cpus` (likely and for other params) if it is default.

Reply:
Using of `Backend` and `CpuCount` configs was added. Lines of code with comment `# TOREVERT` will be moved to part of a separate PR.

Reply:
PR is rebased on current master, extra lines were removed

Reply:
Why is `mp.cpu_count()` still used?

Reply:
It's my omission. Fixed.